Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,484 @@
+            }
+        }
+
+        /// <summary>
+        /// Calculates half frustum height for orthographic or perspective camera.
+        /// For more info on frustum height, see <see cref="docs.unity3d.com/Manual/FrustumSizeAtDistance.html"/> 
+        /// </summary>
+        /// <param name="state">CameraState for checking if Orthographic or Perspective</param>
+        /// <param name="vcam">vcam, to check its position</param>
+        /// <returns>Frustum height of the camera</returns>
+        private float CalculateHalfFrustumHeight(in CameraState state, in float cameraPosLocalZ)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = state.Lens.OrthographicSize;
+            }
+            else
+            {
+                // distance between the collider's plane and the camera
+                float distance = cameraPosLocalZ;
+                frustumHeight = distance * Mathf.Tan(state.Lens.FieldOfView * 0.5f * Mathf.Deg2Rad);
+            }
+
+            return Mathf.Abs(frustumHeight);
+        }
+        
+        /// <summary>
+        /// Confines input 2D point within the confined area.
+        /// </summary>
+        /// <param name="positionToConfine">2D point to confine</param>
+        /// <returns>Confined position</returns>
+        private Vector2 ConfinePoint(Vector2 positionToConfine, in List<List<Vector2>> pathCache,
+            in bool hasBone, in float windowWidth, in float windowHeight)
+        {
+            if (ShrinkablePolygon.IsInside(pathCache, positionToConfine))
+            {
+                return positionToConfine;
+            }
+            bool outsideOfOriginal = !ShrinkablePolygon.IsInside(m_shapeCache.m_OriginalPath, positionToConfine);
+
+            Vector2 closest = positionToConfine;
+            float minDistance = float.MaxValue;
+            for (int i = 0; i < pathCache.Count; ++i)
+            {
+                int numPoints = pathCache[i].Count;
+                for (int j = 0; j < numPoints; ++j)
+                {
+                    Vector2 v0 = pathCache[i][j];
+                    Vector2 v = pathCache[i][(j + 1) % numPoints];
+                    Vector2 c = Vector2.Lerp(v0, v, positionToConfine.ClosestPointOnSegment(v0, v));
+                    Vector2 difference = positionToConfine - c;
+                    float distance = Vector2.SqrMagnitude(difference);
+                    if (Mathf.Abs(difference.x) > windowWidth || Mathf.Abs(difference.y) > windowHeight)
+                    {
+                        // penalty for points from which the target is not visible, prefering visibility over proximity
+                        distance += m_confinerBaker.m_sqrPolygonDiagonal; 
+                    }
+#if CINEMACHINE_EXPERIMENTAL_CONFINER2D
+                    bool isValid = outsideOfOriginal || !hasBone || !DoesIntersectOriginal(positionToConfine, c);
```

Comment:
I don't understand what hasBone has to do with this.  Seems to me that this situation can arise whether there is a bone or not.  What am I missing?

Reply:
hasBone can only happen when experimental is on, because when it is off, we stop at just before hasBone can be true.

But, I realized when experimental off, the only case when I need to check if DoesIntersectectOriginal is when I am at the last confiner state. So I say added hasBone here too. I set hasBone to true in this case, if the orthosize is bigger than the last one if it was called because it would have a bone afterwards.

Reply:
if Outside: valid
if Inside, and no bone: valid
if Inside, and has bone and does not intersect original: valid

Reply:
The reason why I only check DoesIntersectOriginal after hasBone is true, is related to the penalty explanation down below. After the object has a bone (or should have one if experimental was on), then the penalty may not exclude points that are unwanted, so I have to exclude them  with DoesIntersectOriginal.

We can discuss it on zoom. I can show you what I mean.